### ISSUE TITLE
Bug: isUnique() on Cake\ORM\RulesChecker only checks the uniqueness of items filtered by beforeFind() on Cake\ORM\Table

### ISSUE THREAD

**Issue by jadelbe418** [cakephp/cakephp#15644]
I started this as [a question on Stack Overflow](https://stackoverflow.com/questions/68184422/restrict-query-results-on-beforefind-table-call-back-without-breaking-buildrul), which has the most detailed explanation of the issue. Although the more I look into it, it seems like this is a bug.

Basically I have logic in the beforeFind() callbacks on the tables of my SAS application that make it so a user can only view items that are at the same location they are at. It looks something like this:

    <?php
    
    class BaseTable extends Cake\ORM\Table
    {
        public function beforeFind(Event $event, Query $query, ArrayObject $options)
        {
            parent::beforeFind($event, $query, $options);
    
            $repositoryTable = $query->getRepository()->getAlias();
    
            $loacationIds = Cake\ORM\TableRegistry::getTableLocator()
                ->get("Locations")
                ->getUserLocationIds();
    
            $query->andWhere([$repositoryTable . ".location_id IN " => $locationIds]);
            
        }
    }

And then we use the following rule on the inventory items model to make sure that the item being entered for a specific client is unique: 

    <?php
    
    class ItemsTable extends BaseTable
    {
        public function buildRules(Cake\ORM\RulesChecker $rules): Cake\ORM\RulesChecker
        {
            $rules->add($rules->isUnique(['id_number', 'client_id'], 'The Number you selected is in use, please use another'));
            return $rules;
        }
    }

The isUnique call seems to grab the the code in the beforeFind() and only examines the uniqueness of the items within the $loacationIds. It does not prevent the addition of items that violate the isUnique rule when they do not exist within the locations specified in the $loacationIds.

The code comment on the isUnique call (Cake\ORM\RulesChecker, line 32) sates the following as the description for the function: "Returns a callable that can be used as a rule for checking the uniqueness of a value in the table."

    /**
     * Returns a callable that can be used as a rule for checking the uniqueness of a value
     * in the table.
     *
     * ### Example:
     *
     * ```
     * $rules->add($rules->isUnique(['email'], 'The email should be unique'));
     * ```
     *
     * @param string[] $fields The list of fields to check for uniqueness.
     * @param string|array|null $message The error message to show in case the rule does not pass. Can
     *   also be an array of options. When an array, the 'message' key can be used to provide a message.
     * @return callable
     */
    public function isUnique(array $fields, $message = null)
    {
        .....
    }

So I feel that it should be safe to assume that these calls should ignore any query manipulation in the model call backs.

The application I am running into this on is running CakePHP 3.8.

**Comment by markstory** [comment]
> The isUnique call seems to grab the the code in the beforeFind() and only examines the uniqueness of the items within the $loacationIds. It does not prevent the addition of items that violate the isUnique rule when they do not exist within the locations specified in the $loacationIds.

This is on purpose as the IsUnique rules use find() to look for records. Not using find() would cause other problems with behaviors and existing user-land code. If you need to not have beforeFind called on your IsUnique rules you'll need to create your own implementation.

**Comment by jadelbe418** [comment]
The explanation on the function call and seems to state that the uniqueness is determined against the table, and not whatever subset of the table is being filtered in the the events triggered against `find()`:

`Returns a callable that can be used as a rule for checking the uniqueness of a value in the table.`

And the documentation does not seem to [mention anything like this](https://book.cakephp.org/3/en/orm/validation.html#creating-unique-field-rules) either.

I guess I am failing to see what the point of `isUnique()` is if it it is only looking at a subsection of data and not the entire contents of the table.

I suppose it may be the the intended use case of narrowing the result set in the find() call back events (`beforeFind()`) is not part of their intended use then? If that is the case it seems there would be some mention of it within the code comment or documentation, noted above. I would be happy to submit a pull request for it, if that is the case.

I will fix my code with a custom `isUnique()` like you mentioned or the solution that was just posted on my [initial question on Stack Overflow](https://stackoverflow.com/questions/68184422/restrict-query-results-on-beforefind-table-call-back-without-breaking-buildrul), if need be. I appreciate the suggestion.

**Comment by markstory** [comment]
> I guess I am failing to see what the point of isUnique() is if it it is only looking at a subsection of data and not the entire contents of the table.

This only seems to be problematic if your table has `beforeFind()` logic like yours.

> And the documentation does not seem to mention anything like this either.

Fair. I'll get the documentation updated.

**Comment by jadelbe418** [comment]
@markstory, 

I went ahead and implemented a custom App\Model\Rule\BuildRule\UniqueRule() rule to fix this in my code where I can directly query all of the records in the database by passing a queryOptions through an argument on the rule:

    $rules->add(new App\Model\Rule\BuildRule\UniqueRule($this),
            'UniqueSerialNumber',
            [
                'errorField' => 'serial_number',
                'message' => 'The Serial Number you selected is in use, please use another',
                'fields' => ['serial_number'],
                'queryOptions' => ['ignoreSiteFilters' => true]
     ]);

And then it is injected into the query within App\Model\Rule\BuildRule\UniqueRule

    $query = $this->MyTable->find('all', $options['queryOptions'])
            ->select($options['fields']);

With something like this in the beforeFind() on the base table to which the rule is applied:

    public function beforeFind(Event $event, Query $query, ArrayObject $options)
    {
        parent::beforeFind($event, $query, $options);

        if ($options->offsetExists('ignoreSiteFilters') === fasle || $options->offsetGet('ignoreSiteFilters') === false)) {
            $query->andWhere([$repositoryTable . ".site_id IN " => $siteIds]);
        }
    }

I more or less had to do my best to attempt to re-write Cake's isUnique() rule in App\Model\Rule\BuildRule\UniqueRule, with the one difference of being able override the query in beforeFind() with the $options array input.

I had this the answer to my [initial question on stack overflow](https://stackoverflow.com/questions/68184422/restrict-query-results-on-beforefind-table-call-back-without-breaking-buildrul) and after having written this implementation I have found that I agree heavily with it heavily:

_"As far as the core is concerned, maybe it would be helpful if the rules would for example accept a callback that allows to modify the query."_

If something like this existed in it would eliminate the need for a custom rule to facilitate this behavior. I feel like this kind of filtering must be pretty common in SAS applications, so it would be a nice ability to have if it is doable in the core.

Figured it may be worth suggesting.
